File: Server/Presence/src/main/java/org/elkoserver/server/presence/GraphDesc.java
package org.elkoserver.server.presence;

import org.elkoserver.foundation.json.JSONMethod;
import org.elkoserver.json.JsonObject;

import java.lang.reflect.InvocationTargetException;

class GraphDesc {
    private String myClassName;
    private String myGraphName;
    private JsonObject myConf;

    @JSONMethod({ "class", "name", "?conf" })
    GraphDesc(String className, String graphName, JsonObject conf) {
        myClassName = className;
        myGraphName = graphName;
        if (conf == null) {
            conf = new JsonObject();
        }
        myConf = conf;
    }

    SocialGraph init(PresenceServer master) {
        try {
            SocialGraph newGraph =
                (SocialGraph) Class.forName(myClassName).getConstructor().newInstance();
            newGraph.init(master, new Domain(myGraphName), myConf);
            return newGraph;
        } catch (ClassNotFoundException e) {
            master.appTrace().errori("class " + myClassName + " not found");
            return null;
        } catch (InstantiationException | IllegalAccessException e) {
            master.appTrace().errorm("unable to instantiate " + myClassName +
                                     ": " + e);
            return null;
        } catch (NoSuchMethodException e) {
            master.appTrace().errorm("class " + myClassName +
                    " does not have a public no-arg constructor: " + e);
            return null;
        } catch (InvocationTargetException e) {
            master.appTrace().errorm("error occurred during instantiation of " + myClassName +
                    ": " + e.getCause());
            return null;
        }
    }
}

File: Server/Presence/src/main/kotlin/org/elkoserver/server/presence/GraphDesc.kt
package org.elkoserver.server.presence

import org.elkoserver.foundation.json.Injector
import org.elkoserver.foundation.json.JsonMethod
import org.elkoserver.json.JsonObject
import org.elkoserver.util.trace.slf4j.Gorgel
import java.lang.reflect.InvocationTargetException

internal class GraphDesc @JsonMethod("class", "name", "?conf") constructor(
        private val myClassName: String,
        private val myGraphName: String,
        conf: JsonObject?) : DomainRegistryUsingObject, InjectorsUsingObject {
    private val myConf: JsonObject = conf ?: JsonObject()

    private lateinit var graphInjectors: Collection<Injector>

    private lateinit var myDomainRegistry: DomainRegistry

    override fun setDomainRegistry(domainRegistry: DomainRegistry) {
        myDomainRegistry = domainRegistry
    }

    override fun setInjectors(injectors: Collection<Injector>) {
        graphInjectors = injectors
    }

    fun init(master: PresenceServer, gorgel: Gorgel, socialGraphGorgel: Gorgel) =
            try {
                (Class.forName(myClassName).getConstructor().newInstance() as SocialGraph).apply {
                    graphInjectors.forEach { it.inject(this) }
                    init(master, socialGraphGorgel, Domain(myGraphName, myDomainRegistry), myConf)
                }
            } catch (e: ClassNotFoundException) {
                gorgel.error("class $myClassName not found")
                null
            } catch (e: InstantiationException) {
                gorgel.error("unable to instantiate $myClassName: $e")
                null
            } catch (e: IllegalAccessException) {
                gorgel.error("unable to instantiate $myClassName: $e")
                null
            } catch (e: NoSuchMethodException) {
                gorgel.error("class $myClassName does not have a public no-arg constructor: $e")
                null
            } catch (e: InvocationTargetException) {
                gorgel.error("error occurred during instantiation of $myClassName: ${e.cause}")
                null
            }
}
